loom/graphics/gfxTexture.cpp

```diff
@@ -285,33 +357,273 @@ TextureInfo *Texture::initFromBytes(utByteArray *bytes)
     }
 
     // Great, stuff real bits!
-    lmLog(gGFXTextureLogGroup, "loaded image bytes - %i x %i at id %i", lat->width, lat->height, id);
+    lmLog(gGFXTextureLogGroup, "loaded image bytes - %i x %i at id %i", lat->width, lat->height, tinfo->id);
 
-    loadImageAsset(lat, id);
+    loadImageAsset(lat, tinfo->id);
 
     dtor(lat);
 
     return tinfo;
 }
 
-void Texture::loadCheckerBoard(TextureID id)
+
+int __stdcall Texture::loadTextureAsync_body(void *param)
 {
-        const int checkerboardSize = 128, checkSize = 8;        
+    bool addToQueue;
+    const char *path = NULL;
 
-        int *checkerboard = (int*)lmAlloc(gGFXTextureAllocator, checkerboardSize*checkerboardSize*4);
+    //remain in a loop here so long as we have notes to process
+    while(true)
+    {
+        addToQueue = true;
+
+        //get the front of the async texture queue to process
+        loom_mutex_lock(Texture::sAsyncQueueMutex);
+        AsyncLoadNote threadNote = Texture::sAsyncLoadQueue.front();
+        Texture::sAsyncLoadQueue.pop_front();
+        loom_mutex_unlock(Texture::sAsyncQueueMutex);
+        path = (!threadNote.path.empty()) ? threadNote.path.c_str() : NULL;
+
+        //make sure we weren't disposed in the meantime... if so, just skip!
+        loom_mutex_lock(Texture::sTexInfoLock);
+        bool alreadyDisposed = threadNote.tinfo->asyncDispose;
+        if(alreadyDisposed)
+        {
+            //invalidate the TextureInfo and make it available for use again
+            if (!threadNote.tinfo->texturePath.empty()) 
+            {
+                sTexturePathLookup.erase(threadNote.tinfo->texturePath);
+            }
+            threadNote.tinfo->reset();
+            loom_mutex_unlock(Texture::sTexInfoLock);
 
-        for (int i = 0; i < checkerboardSize; i++)
+            //prep for below
+            loom_mutex_lock(Texture::sAsyncQueueMutex);
+        }
+        else
         {
-            for (int j = 0; j < checkerboardSize; j++)
+            loom_mutex_unlock(Texture::sTexInfoLock);
+            
+            //handle Asset vs ByteArray texture load
+            if(path)
+            {
+                // Load async since we're in a background thread.
+                lmLog(gGFXTextureLogGroup, "loading %s async...", path);
+                loom_asset_preload(path);
+                loom_thread_yield();
+                loom_asset_image *lai = NULL;
+                while((lai = (loom_asset_image *)loom_asset_lock(path, LATImage, 0)) == NULL)
```

Comment:
Should this have a failsafe? Is it any different than the blocking lock in this usage?